File: leetcode/104.maximum-depth-of-binary-tree/go/maxdepth.go
package maxdepth

import "container/list"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	queue := list.New()
	depth := 0

	if root != nil {
		queue.PushBack(root)
	}

	for queue.Len() > 0 {
		size := queue.Len()
		depth++

		for range size {
			curr := queue.Remove(queue.Front()).(*TreeNode)

			if curr.Left != nil {
				queue.PushBack(curr.Left)
			}

			if curr.Right != nil {
				queue.PushBack(curr.Right)
			}
		}
	}

	return depth
}

File: leetcode/104.maximum-depth-of-binary-tree/ts/maxdepth.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function maxDepth(root: TreeNode | null): number {
    const queue: TreeNode[] = []
    let front = 0
    let depth = 0

    if (root !== null) queue.push(root)

    while (queue.length > front) {
        const size = queue.length - front
        depth++

        for (let i = 0; i < size; i++) {
            const curr = queue[front++]

            if (curr.left !== null) queue.push(curr.left)
            if (curr.right !== null) queue.push(curr.right)
        }
    }

    return depth
}

export default maxDepth
